[SPARK-32907][ML] adaptively blockify instances - LinearSVC

zhengruifeng opened the pull request:
### What changes were proposed in this pull request?
1, use `blockSizeInMB` instead of `blockSize`(#rows) to control the stacking of vectors;
2, infer an appropriate `blockSizeInMB` if set 0, based on data sparsity;

### Why are the changes needed?
the performance gain is mainly related to the nnz of block.


### Does this PR introduce _any_ user-facing change?
yes, param `blockSize` -> `blockSizeInMB` in master


### How was this patch tested?
added testsuites and performance test (result attached in [ticket](https://issues.apache.org/jira/browse/SPARK-32907))

zhengruifeng committed:
init

SparkQA commented:
**[Test build #129613 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129613/testReport)** for PR 29998 at commit [`a2a4f45`](https://github.com/apache/spark/commit/a2a4f454d9a52f1970c3f5b9909996ffc685079f).

zero323 committed:
[SPARK-33105][INFRA] Change default R arch from i386 to x64 and parametrize BINPREF

### What changes were proposed in this pull request?

- Change default R `arch` from `i386` to `x64`, to match Rtools  version.
- Parameterize `BINPREF` with `WIN` (https://stackoverflow.com/a/44035904)

Reported on dev:

http://apache-spark-developers-list.1001551.n3.nabble.com/Broken-rlang-installation-on-AppVeyor-td30294.html

### Why are the changes needed?

It seems like update from rlang 0.4.7 to 0.4.8 exposed an issue, where build fails because of incompatible ddl

```
c:/Rtools40/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe:
skipping incompatible C:/R/bin/i386/R.dll when searching for -lR
[00:01:52]
c:/Rtools40/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe:
skipping incompatible C:/R/bin/i386/R.dll when searching for -lR
[00:01:52]
c:/Rtools40/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe:
cannot find -lR
[00:01:52] collect2.exe: error: ld returned 1 exit status
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing tests.

Closes #29991 from zero323/APPVEYOR-DEAFAULT-ARCH.

Authored-by: zero323 <mszymkiewicz@gmail.com>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

gaborgsomogyi committed:
[SPARK-33102][SQL] Use stringToSeq on SQL list typed parameters

### What changes were proposed in this pull request?
While I've implemented JDBC provider disable functionality it has been popped up [here](https://github.com/apache/spark/pull/29964#discussion_r501786746) that `Utils.stringToSeq` must be used when String list type SQL parameter handled. In this PR I've fixed the problematic parameters.

### Why are the changes needed?
`Utils.stringToSeq` must be used when String list type SQL parameter handled.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existing unit tests.

Closes #29989 from gaborgsomogyi/SPARK-33102.

Authored-by: Gabor Somogyi <gabor.g.somogyi@gmail.com>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

SparkQA commented:
**[Test build #129613 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129613/testReport)** for PR 29998 at commit [`a2a4f45`](https://github.com/apache/spark/commit/a2a4f454d9a52f1970c3f5b9909996ffc685079f).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait HasBlockSizeInMB extends Params `
  * `class HasBlockSizeInMB(Params):`

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34216/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129613/
Test FAILed.

zhengruifeng committed:
fix py

SparkQA commented:
**[Test build #129615 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129615/testReport)** for PR 29998 at commit [`668778c`](https://github.com/apache/spark/commit/668778cb0b090b4b6e815479f76821b04d3c56ea).

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34216/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34216/
Test PASSed.

dongjoon-hyun committed:
[SPARK-33108][BUILD] Remove sbt-dependency-graph SBT plugin

### What changes were proposed in this pull request?

This PR aims to remove `sbt-dependency-graph` SBT plugin.

### Why are the changes needed?

`sbt-dependency-graph` officially doesn't support SBT 1.3.x and it's broken due to `NoSuchMethodError`. This cannot be fixed in `sbt-dependency-graph` side at SBT 1.3.x
- https://github.com/sbt/sbt-dependency-graph
    > Note: Under sbt >= 1.3.x some features might currently not work as expected or not at all (like dependencyLicenses).

```
$ build/sbt dependencyTree
Launching sbt from build/sbt-launch-1.3.13.jar
[info] welcome to sbt 1.3.13 (AdoptOpenJDK Java 1.8.0_252)
...
[error] java.lang.NoSuchMethodError: sbt.internal.LibraryManagement$.cachedUpdate(Lsbt/librarymanagement/DependencyResolution;Lsbt/librarymanagement/ModuleDescriptor;Lsbt/util/CacheStoreFactory;Ljava/lang/String;Lsbt/librarymanagement/UpdateConfiguration;Lscala/Function1;ZZZLsbt/librarymanagement/UnresolvedWarningConfiguration;Lsbt/librarymanagement/EvictionWarningOptions;ZLsbt/internal/librarymanagement/CompatibilityWarningOptions;Lsbt/util/Logger;)Lsbt/librarymanagement/UpdateReport;
```

**ALTERNATIVES**
- One alternative is `coursier`, but it requires `coursier-based sbt launcher` which is more intrusive.
  - https://get-coursier.io/docs/sbt-coursier.html#sbt-13x
    > you'll have to use the coursier-based sbt launcher, via its custom sbt-extras launcher for example.

- Another alternative is moving to `SBT 1.4.0` which uses `sbt-dependency-graph` as a built-in, but it's still new and will requires many change.

So, this PR aims to remove the broken plugin simply.

### Does this PR introduce _any_ user-facing change?

No. This is a dev-only change.

### How was this patch tested?

Manual.
```
$ build/sbt dependencyTree
...
[error] Not a valid command: dependencyTree
[error] Not a valid project ID: dependencyTree
[error] Not a valid key: dependencyTree (similar: dependencyOverrides, sbtDependency, dependencyResolution)
[error] dependencyTree
[error]               ^
```

Closes #29997 from dongjoon-hyun/remove_depedencyTree.

Lead-authored-by: Dongjoon Hyun <dongjoon@apache.org>
Co-authored-by: Dongjoon Hyun <dhyun@apple.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34218/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34218/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34218/
Test PASSed.

SparkQA commented:
**[Test build #129615 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129615/testReport)** for PR 29998 at commit [`668778c`](https://github.com/apache/spark/commit/668778cb0b090b4b6e815479f76821b04d3c56ea).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129615/
Test PASSed.

zero323 reviewed (commented):
We should also have modify Python type annotations:

- Adjust type definitiion and `__init__`  of [`_LinearSVCParams`](https://github.com/apache/spark/blob/dfb7790a9dad8e98bd27001a613b4e13a5eb9d51/python/pyspark/ml/classification.pyi#L165-L178)
- Adjust `__init__` and `setParams` of [`LinearSVC`](https://github.com/apache/spark/blob/dfb7790a9dad8e98bd27001a613b4e13a5eb9d51/python/pyspark/ml/classification.pyi#L186-L219)
- Add annotations for new method and class.

zhengruifeng committed:
adjust classification.pyi

SparkQA commented:
**[Test build #129619 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129619/testReport)** for PR 29998 at commit [`4fba526`](https://github.com/apache/spark/commit/4fba5264b2c7b8d0da107867e6f218514b3e819a).

zhengruifeng committed:
update suite

SparkQA commented:
**[Test build #129620 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129620/testReport)** for PR 29998 at commit [`f756d9c`](https://github.com/apache/spark/commit/f756d9cbf6764ea51143fd928ad5a1a1da985426).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34223/

SparkQA commented:
**[Test build #129619 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129619/testReport)** for PR 29998 at commit [`4fba526`](https://github.com/apache/spark/commit/4fba5264b2c7b8d0da107867e6f218514b3e819a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34223/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34223/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34224/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/34224/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/34224/
Test PASSed.

SparkQA commented:
**[Test build #129620 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/129620/testReport)** for PR 29998 at commit [`f756d9c`](https://github.com/apache/spark/commit/f756d9cbf6764ea51143fd928ad5a1a1da985426).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/129620/
Test PASSed.

zhengruifeng commented:
also ping @WeichenXu123

zero323 commented:
Thanks for the changes @zhengruifeng 

Could also modify `shared.pyi` and add the import?

```
diff --git a/python/pyspark/ml/classification.pyi b/python/pyspark/ml/classification.pyi
index 74d11cfd11..241f5baf8d 100644
--- a/python/pyspark/ml/classification.pyi
+++ b/python/pyspark/ml/classification.pyi
@@ -26,6 +26,7 @@ from pyspark.ml.base import _PredictorParams
 from pyspark.ml.param.shared import (
     HasAggregationDepth,
     HasBlockSize,
+    HasBlockSizeInMB,
     HasElasticNetParam,
     HasFitIntercept,
     HasMaxIter,
diff --git a/python/pyspark/ml/param/shared.pyi b/python/pyspark/ml/param/shared.pyi
index 5999c0eaa4..bbb4890455 100644
--- a/python/pyspark/ml/param/shared.pyi
+++ b/python/pyspark/ml/param/shared.pyi
@@ -185,3 +185,8 @@ class HasBlockSize(Params):
     blockSize: Param[int]
     def __init__(self) -> None: ...
     def getBlockSize(self) -> int: ...
+
+class HasBlockSizeInMB(Params):
+    blockSizeInMB: Param[float]
+    def __init__(self) -> None: ...
+    def getBlockSizeInMB(self) -> float: ...
```

In general you can verify consistency of annotations locally (sorry, still working on the pipeline):

```bash
# Checks annotations
mypy --no-incremental --config python/mypy.ini python/pyspark
# Checks Spark examples
MYPYPATH=python/ mypy --no-incremental --config python/mypy.ini examples/src/main/python/ml examples/src/main/python/sql examples/src/main/python/sql/streaming
```

holdenk committed:
[SPARK-32881][CORE] Catch some race condition errors and log them more clearly

### What changes were proposed in this pull request?

Decommissioning can run out of time resulting in some race condition, these race conditions result in confusing error messages but not negative impact.

### Why are the changes needed?

The NPE & element missing errors in the log can create a missunderstanding.

### Does this PR introduce _any_ user-facing change?
Logs change.

### How was this patch tested?
Existing tests pass.

Closes #29992 from holdenk/SPARK-32881-error-messaging-on-decom-race-messages.

Authored-by: Holden Karau <hkarau@apple.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

wangyum committed:
[SPARK-33107][SQL] Remove hive-2.3 workaround code

### What changes were proposed in this pull request?

This pr remove `hive-2.3` workaround code.

### Why are the changes needed?

Make code more clear and readable.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing unit tests.

Closes #29996 from wangyum/SPARK-33107.

Authored-by: Yuming Wang <yumwang@ebay.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

zero323 committed:
[SPARK-33086][FOLLOW-UP] Remove unused Optional import from pyspark.resource.profile stub

### What changes were proposed in this pull request?

Remove unused `typing.Optional` import from `pyspark.resource.profile` stub.

### Why are the changes needed?

Since SPARK-32319 we don't allow unused imports.  However, this one slipped both local and CI tests for some reason.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests and mypy check.

Closes #30002 from zero323/SPARK-33086-FOLLOWUP.

Authored-by: zero323 <mszymkiewicz@gmail.com>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

gemelen committed:
[SPARK-33106][BUILD] Fix resolvers clash in SBT

### What changes were proposed in this pull request?

Rename manually added resolver for local Ivy repo.
Create configuration to publish to local Ivy repo similar to Maven one.
Use `publishLocal` to publish both to local Maven and Ivy repos instead
of custom task `localPublish` (renamed from `publish-local` of sbt 0.13.x).

### Why are the changes needed?

There are two resolvers (bootResolvers's "local" and manually added "local") that point to the same local Ivy repo, but have different configurations, which led to excessive warnings in logs and, potentially, resolution issues.
Changeset fixes that case, observable in sbt output as
```
[warn] Multiple resolvers having different access mechanism configured with same name 'local'. To avoid conflict, Remove duplicate project resolvers (`resolvers`) or rename publishing resolve
r (`publishTo`).
```

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Executed `build/sbt`'s `publishLocal` task on individual module and on root project.

Closes #30006 from gemelen/feature/local-resolvers.

Authored-by: Denis Pyshev <git@gemelen.net>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

zhengruifeng committed:
update pyi

zhengruifeng commented:
@zero323 Thank you so much. I involved other commits here by mistake. I will resend a new PR.

WeichenXu123 commented on project/plugins.sbt:
```diff
@@ -27,8 +27,6 @@ addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "0.15.0")
 addSbtPlugin("com.typesafe.sbteclipse" % "sbteclipse-plugin" % "5.2.4")
 
-addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.9.2")
-
```
why change this ?

WeichenXu123 commented on project/SparkBuild.scala:
```diff
@@ -213,17 +213,23 @@ object SparkBuild extends PomBuild {
       Resolver.mavenLocal,
-      Resolver.file("local", file(Path.userHome.absolutePath + "/.ivy2/local"))(Resolver.ivyStylePatterns)
+      Resolver.file("ivyLocal", file(Path.userHome.absolutePath + "/.ivy2/local"))(Resolver.ivyStylePatterns)
     ),
     externalResolvers := resolvers.value,
     otherResolvers := SbtPomKeys.mvnLocalRepository(dotM2 => Seq(Resolver.file("dotM2", dotM2))).value,
     publishLocalConfiguration in MavenCompile := PublishConfiguration()
         .withResolverName("dotM2")
         .withArtifacts(packagedArtifacts.value.toVector)
         .withLogging(ivyLoggingLevel.value),
+    publishLocalConfiguration in SbtCompile := PublishConfiguration()
+        .withResolverName("ivyLocal")
+        .withArtifacts(packagedArtifacts.value.toVector)
+        .withLogging(ivyLoggingLevel.value),
     publishMavenStyle in MavenCompile := true,
+    publishMavenStyle in SbtCompile := false,
     publishLocal in MavenCompile := publishTask(publishLocalConfiguration in MavenCompile).value,
-    publishLocalBoth := Seq(publishLocal in MavenCompile, publishLocal).dependOn.value,
+    publishLocal in SbtCompile := publishTask(publishLocalConfiguration in SbtCompile).value,
+    publishLocal := Seq(publishLocal in MavenCompile, publishLocal in SbtCompile).dependOn.value,
```
why change this ?

WeichenXu123 commented on python/pyspark/resource/profile.pyi:
```diff
@@ -22,7 +22,7 @@ from pyspark.resource.requests import (  # noqa: F401
     TaskResourceRequest as TaskResourceRequest,
     TaskResourceRequests as TaskResourceRequests,
 )
-from typing import overload, Dict, Optional, Union
```
why change this ?

WeichenXu123 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala:
```diff
@@ -76,7 +76,7 @@ class MicroBatchExecution(
 
     val disabledSources =
-      sparkSession.sqlContext.conf.disabledV2StreamingMicroBatchReaders.split(",")
+      Utils.stringToSeq(sparkSession.sqlContext.conf.disabledV2StreamingMicroBatchReaders)
```
why change this ?

WeichenXu123 reviewed (commented):
This PR includes many irrelevant changes ?
